cherab/core/model/plasma/thermal_cx.pyx

```diff
@@ -0,0 +1,159 @@
+    :param AtomicData atomic_data: The atomic data provider for this model. Default is None.
+    :param object lineshape: Line shape model class. Default is None (GaussianLine).
+    :param object lineshape_args: A list of line shape model arguments. Default is None.
+    :param object lineshape_kwargs: A dictionary of line shape model keyword arguments. Default is None.
+
+    :ivar Plasma plasma: The plasma to which this emission model is attached.
+    :ivar AtomicData atomic_data: The atomic data provider for this model.
+    """
+
+    def __init__(self, Line line, Plasma plasma=None, AtomicData atomic_data=None, object lineshape=None,
+                 object lineshape_args=None, object lineshape_kwargs=None):
+
+        super().__init__(plasma, atomic_data)
+
+        self._line = line
+
+        self._lineshape_class = lineshape or GaussianLine
+        if not issubclass(self._lineshape_class, LineShapeModel):
+            raise TypeError("The attribute lineshape must be a subclass of LineShapeModel.")
+
+        if lineshape_args:
+            self._lineshape_args = lineshape_args
+        else:
+            self._lineshape_args = []
+        if lineshape_kwargs:
+            self._lineshape_kwargs = lineshape_kwargs
+        else:
+            self._lineshape_kwargs = {}
+
+        # ensure that cache is initialised
+        self._change()
+
+    def __repr__(self):
+        return '<ThermalCXLine: element={}, charge={}, transition={}>'.format(self._line.element.name, self._line.charge, self._line.transition)
+
+    cpdef Spectrum emission(self, Point3D point, Vector3D direction, Spectrum spectrum):
+
+        cdef:
+            double ne, te, receiver_density, donor_density, donor_temperature, weighted_rate, radiance
+            Species species
+            ThermalCXPEC rate
+
+        # cache data on first run
+        if self._target_species is None:
+            self._populate_cache()
+
+        ne = self._plasma.get_electron_distribution().density(point.x, point.y, point.z)
+        if ne <= 0.0:
+            return spectrum
+
+        te = self._plasma.get_electron_distribution().effective_temperature(point.x, point.y, point.z)
+        if te <= 0.0:
+            return spectrum
+
+        receiver_density = self._target_species.distribution.density(point.x, point.y, point.z)
+        if receiver_density <= 0.0:
+            return spectrum
+
+        weighted_rate = 0
+        for species, rate in self._rates:
+            donor_density = species.distribution.density(point.x, point.y, point.z)
+            donor_temperature = species.distribution.effective_temperature(point.x, point.y, point.z)
+            weighted_rate += donor_density * rate.evaluate(ne, te, donor_temperature)
+
+        # add emission line to spectrum
+        radiance = RECIP_4_PI * weighted_rate * receiver_density
+        return self._lineshape.add_line(radiance, point, direction, spectrum)
+
+    cdef int _populate_cache(self) except -1:
```

Comment:
This is consistent with how `ExcitationLine` and `RecombinationLine` define their `_populate_cache` methods. I think it stems from the following in the [Cython documentation](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values):

> There’s a known performance pitfall when combining nogil and except * @cython.exceptval(check=True). In this case Cython must always briefly re-acquire the GIL after a function call to check if an exception has been raised. This can commonly happen with a function returning nothing (C void). Simple workarounds are to mark the function as noexcept if you’re certain that exceptions cannot be thrown, or to change the return type to int and just let Cython use the return value as an error flag (by default, -1 triggers the exception check).

That doesn't actually apply here (the function isn't `nogil`), and `_populate_cache` isn't going to be called often enough for any performance penalty to matter anyway. My guess is that it's a relic of either some older organisation of the code where this did matter but no longer applies, or it's a bit of premature optimisation. Either way I don't think it matters, and I have no qualms with following the existing convention as it's harmless.

Reply:
As @jacklovell  wrote, this is an integer function only to maintain consistency with other models in Cherab. This doesn't affect anything.